### ISSUE TITLE
Use multiple JSON files for one environment

### ISSUE THREAD

**Issue by drj613** [mockoon/mockoon#877]
It would be helpful to have multiple JSON files for one environment - potentially a different JSON file for every endpoint in that env? I'm working on a new application that has multiple people creating new endpoints and we've found that resolving merge conflicts for the `api-mock.json` file can be difficult.

In my head this feature looks something like:
- point mockoon app to a `mockoon` directory
- inside `/mockoon` have an `index.json` file and several endpoint files (eg `my-endpoint-GET.json`)
- app will use `index.json` just for endpoint names and filepaths
- any changes made to an endpoint will touch only that specific endpoint's file
- my changing `endpoint-xyz` will no longer cause a merge conflict with Jane Doe when she tries to make changes to `endpoint-abc`

I haven't gone digging through the source code so I don't know if this is so far off from the existing paradigms that it's not worth the effort, but it could be a nice QoL feature

**Comment by UmeshG-24099** [comment]
Yes, if you want to change more than one **route response of different endpoints**, it is advisable to use a data bucket for that specific environment. As you know data bucket is an **environment specific** property and every time we need to restart it after making changes

**Comment by iFederx** [comment]
I also wanted to suggest the same idea!
I spend a lot of time solving conflicts. Having one file per endpoint would also be the solution in my opinion :)

**Comment by vltansky** [comment]
That would save us from huge merge conflicts

**Comment by akashvekariya** [comment]
Yes we really need this feature. It makes management more easier and mockoon the best open-source mock API tool

**Comment by vltansky** [comment]
@255kb I'd be happy to contribute if this feature will be accepted and some guidance.

**Comment by 255kb** [comment]
@vltansky Sure, I would be happy to get some help. 
This is quite a big change to be honest, but I think it could be mitigated by splitting the file just before saving. 
It wouldn't solve the conflicts issue inside a specific route (e.g. two persons adding a route response) but I think it would be a good step forward. 
We can discuss the technicalities on our [Discord server](https://discord.com/invite/JmTBk59MTb) if you want.

**Comment by edtoken** [comment]
@vltansky @255kb guys, I would say that I do not really agree with the potential solution in the MR, let me explain.

The problem is not with routes, the problem is that it's impossible to resolve conflicts because of the environment file structure, and how the diff looks in pull requests.

Where is the problem?
- route
- route responses
- route response body
- data buckets
- data bucket value
- A LOT OTHER PROPERTIES, 

So, this is not only about routes, this is about the whole configuration file.

I would suggest a little another view of the solution.

I just saying how to save configuration in a way that allows you to resolve pull request conflicts, and this already works for me.

So, it's just technical things that already implemented by me locally and works as isolated tool, but could be great if we will merge our ideas and have great integrated feature in a application.

What I have
- extract script
- build script

The extract script is a super easy script that go by keys like this

```
routes: routes.map(processExtract)
```
Almost each property could be processed with same method, except a few special cases.

The extract script splits the original configuration file into chunks that are stored in the user-friendly structure, and each key of the configuration file is processed with a base processor OR a custom processor that makes a structure/nested extract if needed.

Also, this script returns the original configuration, but, the big JSON pieces are replaced with the file path where this piece stored is. 

This allows the build script (opposite flow) to just go through the file and replace file paths in the small config with the extracted data by path, super easy! No any array conflicts/etc.


The extract script:
```
{
"data": [data, data, data]
"routes": [ route, route, route],
}
->
{
data: [filepath, filepath, filepath], 
routes: [filePath, filePath, filePath],
...
}
```

The build script:
```
{
data: [filepath, filepath, filepath], 
routes: [filePath, filePath, filePath],
...
}
->
{
"data": [data, data, data]
"routes": [ route, route, route],
}
```
Easy!


Where is the `processExtract` method 99% same for all properties and makes structure like this 
```
root/index.json
data/bucketname/index.json
routes/http/get/escaped/route/path/index.json
routes/http/get/escaped/route/path/response.200.uuid.json
routes/http/get/escaped/route/path/response.401.uuid.json
```

This allows the extract script to make as user-friendly a structure as we need for pull requests.

And, the build script really does not worry about that structure, nothing complicated there, just goes through the array and restores the data.

This structure will not have conflicts if you 
- updated data backed value
- updated route body 
- updated route uri
- updated route rules
- updated anything else

You will see the normal diff in the expected structure.

So, why save routes as files will not really work and you have to use a little complex system (but a really easy script)

- The filepath implemented not in safely and user friendly way dirs, imagine route like this `/LKJ(*&^*&#[{}*^!@#/1/123/123/{P*&!@#`, I know, looks strange, but possible
- The route could not be represented as a single file, this will not help with conflicts issue because of responses and route body
- Response body not formatted and still one line string (not help with conflicts, the part that mostly updated)
- Data bucket not extracted
- Data value not formatted (same thing as a route body)
- Router details like responses, are not extracted, so, really, you will have conflicts
- Other properties that also could be updated not extracted (headers, settings, etc)


You will definitely have conflicts that are hard to resolve because of these things

My tool works with no updates of Mockoon, as an independent script :) but if we have some compromises and combine whole things into the Mockoon would be great.

**Comment by edtoken** [comment]
i'm not saying that the implementation should be in 2 scripts, I'm saying that the json structure should be implemented in a little another flow

**Comment by edtoken** [comment]
Guys @drj613  @iFederx  @akashvekariya @vltansky I did a small CLI tool that covers all these issues [mockoon-config-transformer](https://github.com/edtoken/mockoon-config-transformer) and a little more.

If someone will work on that feature inside a Mockoon please take a look at this tool, because some specific edge cases are covered there and could be missed by mistake. Also, I will be really glad to be a reviewer or help with that functionality inside a Mockoon, because we really need this :)

**Comment by vltansky** [comment]
hey @edtoken, tried your CLI - looks really good so far, the only issue I had till now is `:` (colon) in filenames not supported on Windows. Will need to pay attention to this in mockoon implementation.

**Comment by edtoken** [comment]
@vltansky thanks! Yeah, thank you! I will fix the dirs/files escaping and release the next version next days.

Really if at some time something like this will be a part of Mockoon then we should pay attention to a bunch of things (I've tried to cover whole things in my tool)

**Technical requirements**
- max filename length/max directory name length
- when you cut filename/directory/path because of OS limit at some time duplicated names could be generated by accident
- special chars that are supported in the route URI could not supported as directory/filename
- routes could not include `/` at all, some examples that were published as Mockoon samples included only hashed URIs with special chars.
- duplicated routes/data bucket/etc names.
- environment structure updates
- friendly and expected (not uuid) folders/files naming
- etc.

For now, I hope my CLI can help with that issue, I will push some requested changes next days.

@vltansky I know that you probably already working on something like this as a Mockoon part, if you need any help or have any updates let me know, please, I will be glad to help/review/etc :)

**Comment by markovalenko** [comment]
@edtoken thank you a lot!

**Comment by edtoken** [comment]
@vltansky updated!

ok, it's time to back to my Mockoon pull requests :)

**Comment by BenHesketh21** [comment]
This would also be useful for having a base JSON and being able to add to it. For us we have an application used by multiple tenants and we need to mock different responses depending on the tenant, yes we can have multiple repsonses but when we have 10 tenants and a change is needed to a common route between them, it means making the same change 10 times. If we could have a JSON with with the common routes and then combine them with tenant specific routes that would be really useful.

**Comment by iFederx** [comment]
Is there any issue with the PR?
@edtoken @vltansky

**Comment by 255kb** [comment]
@iFederx The PR is currently incomplete as it only splits each route into a separate file, but it doesn't prevent most merge conflicts. @edtoken worked on a separate tool (see comments above) that is more comprehensive. However, I'm still unsure if it preserves the full setup (routes order, etc.).

On a side note, I'm a bit concerned about the overhead of such a solution. There is certainly a tradeoff in keeping everything in one file, but I think the simplicity of it is beneficial in many use cases. Splitting into many files will make things more complex at all levels. Additionally, the app wasn't architected this way at all, so integrating this properly could be a headache. 
I can see it becoming a plugin (when and if we add a plugin system), though.

**Comment by Hypercubed** [comment]
Just stumbled across mockoon and this feature request.  Wondering if having a "full match path" helper that gets the full matched path (or some representation of it) could help here.  Right now mockoon can server static files using a mock endpoint (see https://mockoon.com/tutorials/create-endpoint-serving-static-file/) but this doesn't work well for deep paths or paramaterized paths.

As an example I can create a mock endpoint with:

Path: `backend/:path1/:path2`
"File" response body type: `./mock-backed/{{urlParam "path1"}}/{{urlParam "path2"}}/{{ lowercase (method) }}.json`

but this only works for two url segments.

If I had a `{{path}}` token we could do something like this:

Path: `backend/**`
"File" response body type: `./mock-backed/{{ path }}/{{ lowercase (method) }}.json`

for example.

**Comment by serbrech** [comment]
Just providing my 2c on this issue as I would also appreciate better export format to help with conflicts.

- smaller files definitely help, but deep folder structure does not IMO. 1 folder with many files would be simpler.
- automatically naming files is interesting, but the complexity of the solution and number of edgecases isn't appealing. Can we use endpoint name + index instead? or route name if applicable?

I think by simplifying the solution with these 2 trade-offs, it might be simple enough to consider integrating in mockoon?

**Comment by itstueben** [comment]
@255kb Is it possible to add the file feature(like responses) for data buckets? I think this could be a first step. Then the PRs/commits a getting a bit more readable if using git to manage the mockoon file

**Comment by 255kb** [comment]
@itstueben This is not planned yet. 

Sorry that I let this issue without answer for a long time, it is definitely a popular one. The truth is, I have mixed feelings about it. 
I really like the single file model which is easy to use with the CLI or Docker, works well with our cloud, etc. 
I also see the tremendeous value of having a git friendly storage model that makes collaboration easier. 

Now, because I love open-source and I want to be as transparent as possible, this feature also directly competes with the "real-time collaboration" part of our cloud offering, which solves this specific pain point. 
That sounds probably lame and I apologize. To add a bit of context, we are a very small team working on the cloud, not everybody is full-time, and I only created the cloud offering to finance the work (mostly my work) being done on the open-source project. Even if the cloud is showing good signs, we are not VC backed and not drowning in money yet. My priority right now is ensuring the project remains financially sustainable so it doesn't disappear.
I'm keeping this open because I'd love to find a way to make this work in the future without jeopardizing the project's survival, but for now, it's not something I can or will prioritize.
